signer/apk2/apk2.go

```diff
@@ -0,0 +1,148 @@
+package apk2
+
+import (
+	"fmt"
+	"io/ioutil"
+
+	"crypto/ecdsa"
+	"crypto/sha256"
+	"crypto/x509"
+	"os"
+	"os/exec"
+
+	"github.com/pkg/errors"
+	log "github.com/sirupsen/logrus"
+	"go.mozilla.org/autograph/signer"
+)
+
+const (
+	// Type of this signer is "apk2" represents a signer that
+	// shells out to apksigner to sign artifacts
+	Type = "apk2"
+)
+
+// APK2Signer holds the configuration of the signer
+type APK2Signer struct {
+	signer.Configuration
+	minSdkVersion string
+	pkcs8Key      []byte
+}
+
+// New initializes an apk signer using a configuration
+func New(conf signer.Configuration) (s *APK2Signer, err error) {
+	s = new(APK2Signer)
+
+	if conf.Type != Type {
+		return nil, errors.Errorf("apk2: invalid type %q, must be %q", conf.Type, Type)
+	}
+	s.Type = conf.Type
+
+	if conf.ID == "" {
+		return nil, errors.New("apk2: missing signer ID in signer configuration")
+	}
+	s.ID = conf.ID
+
+	if conf.PrivateKey == "" {
+		return nil, errors.New("apk2: missing private key in signer configuration")
+	}
+	s.PrivateKey = conf.PrivateKey
+	priv, err := conf.GetPrivateKey()
+	if err != nil {
+		return nil, errors.Wrap(err, "apk2: failed to get private key from configuration")
+	}
+	switch priv.(type) {
+	case *ecdsa.PrivateKey:
+		// ecdsa is only supported in sdk 18 and higher
+		s.minSdkVersion = "18"
+	default:
+		s.minSdkVersion = "9"
+	}
```

Comment:
nit: might be useful to log which version we pick

Reply:
signers don't log currently, so while I agree with you, that seems like a bigger problem to solve (and one worth solving)

Reply:
> signers don't log currently

This is news to me and the XPI signer logs.

If there are other constraints or best practices we want to enforce like https://github.com/mozilla-services/autograph/issues/247 it'd be good to document them.

Reply:
yeah, I don't know why I said that. we obviously log in the signers. 🤒

Reply:
np